scripts/ASTImportTest.sh

```diff
@@ -153,6 +168,118 @@ function test_ast_import_export_equivalence
     TESTED=$((TESTED + 1))
 }
 
+function run_solc
+{
+    local parameters=( "${@}" )
+
+    if ! "${SOLC}" "${parameters[@]}" > /dev/null 2> solc_stderr
+    then
+        printError "ERROR: ${parameters[*]}"
+        printError "${PWD}"
+        local SOLC_STDERR
+        SOLC_STDERR=$(cat "solc_stderr")
+        printError "${SOLC_STDERR}"
+        exit 1
+    fi
+    rm -f solc_stderr
+
+    return 0
+}
+
+function run_solc_store_stdout
+{
+    local output_file=$1
+    local parameters=( "${@:2}" )
+
+    if ! "${SOLC}" "${parameters[@]}" > "${output_file}" 2> "${output_file}.error"
+    then
+        printError "ERROR: ${parameters[*]}"
+        printError "${PWD}"
+        local SOLC_STDERR
+        SOLC_STDERR=$(cat "${output_file}.error")
+        printError "${SOLC_STDERR}"
+        exit 1
+    fi
+    rm -f "${output_file}.error"
+
+    return 0
+}
+
+function test_evmjson_import_export_equivalence
+{
+    local sol_file="$1"
+    local input_files=( "${@:2}" )
+    local export_command
+    local import_command
+
+    # Compile files (generates the C.bin)
+    mkdir -p sol
+    run_solc --bin "${input_files[@]}" --output-dir sol/
+    for file in sol/*
+    do
+        # If file is empty, delete it.
+        [[ -s "${file}" ]] || rm -f "${file}"
+    done
+
+    mkdir -p asm_json
+    export_command=("${SOLC}" --asm-json "${input_files[@]}" --output-dir asm_json/)
+
+    # Generate asm json file (generates C_evm.json)
+    run_solc --asm-json "${input_files[@]}" --output-dir asm_json/
+    for file in asm_json/*.json
+    do
+        if [[ -n "$(jq --raw-output '. // empty' "$file")" ]]
+        then
+            # Remove '_evm' part of filename '<contract-name>_evm.json'
+            mv "${file}" "${file/_evm/}"
+        else
+            # If file contains 'null', delete it.
+            rm -f "${file}"
```

Comment:
Maybe it is some regression from https://github.com/ethereum/solidity/pull/8783 or even the desired behavior, i.e. return `null` instead of `{}`.

Reply:
> Maybe it is some regression from #8783 or even the desired behavior, i.e. return null instead of {}.

No, it's not a regression. That PR was only about removing unnecessary dict keys. Well, skipping files that are just `null` would fit right in in there, but IIRC doing that was never considered back then.

> Yeah, I honestly don't really know yet the source of the problem

I mean, `null` is valid JSON and if there's no assembly to output, it makes just as much sense as `{}`, if not more. So it's not invalid in a technical sense, just unnecessarily harder to process than if the file was skipped (by the same logic we skip those empty keys). But changing that now would be breaking I guess, so we have to live with it.

> Note that importing an empty JSON object is also not allowed:

That's actually fine, because I don't think we ever output an empty object. So that should stay as is.

The problem is specifically with `null`. It makes no sense for the compiler not to accept the content it does create. I think it would be best if the import feature just accepted those `null`s and treated them as empty assemblies.

I'd request this as a change, but I don't want to block merging the PR just because of it. It's fine done separately. It won't be a common case anyway, it's just an annoying quirk that makes scripting harder for no reason.

Reply:
No, that's probably non-empty. It does not generate that `null` when you export it to JSON, does it?

It should rather match the behavior of something like `echo 'interface I {}' | solc -`.

Reply:
But maybe we need to not use `StrictModeCharReaderBuilder` which seems to ignore `null` and use `StreamWriterBuilder` instead passing `"allowDroppedNullPlaceholders": true` I guess: https://open-source-parsers.github.io/jsoncpp-docs/doxygen/class_json_1_1_stream_writer_builder.html#a79bdf2e639a52f4e758c0b95bd1d3423 the default is `false`: https://open-source-parsers.github.io/jsoncpp-docs/doxygen/class_json_1_1_stream_writer_builder.html#a53bf106b141e28637b01ad0ecd2acbf6

Reply:
> echo 'interface I {}' | solc -

I don't get it. What is the empty assembly in this case? Like, nothing is generated by the compiler, so there is no assembly there. And `echo '' | solc --strict-assembly -` is a invalid assembly input, because a object is expected.

Reply:
> But maybe we need to change the use of `StrictModeCharReaderBuilder` which seems to consider `null` values and use `StreamWriterBuilder` instead passing `"allowDroppedNullPlaceholders": true` I guess: https://open-source-parsers.github.io/jsoncpp-docs/doxygen/class_json_1_1_stream_writer_builder.html#a79bdf2e639a52f4e758c0b95bd1d3423.
> 
> ```
> "dropNullPlaceholders": false or true
> Drop the "null" string from the writer's output for nullValues. 
> Strictly speaking, this is not valid JSON.
> But when the output is being fed to a browser's JavaScript, 
> it makes for smaller output and the browser can handle the output just fine.
> ```
> 
> The default is `false`: https://open-source-parsers.github.io/jsoncpp-docs/doxygen/class_json_1_1_stream_writer_builder.html#a53bf106b141e28637b01ad0ecd2acbf6

What I mean is that using the `jsonParseStrict()` currently requires the input JSON to be either an array or an object value, and apparently `null` is not, or I'm missing something.

Reply:
Ok, I did an example here in this commit: https://github.com/ethereum/solidity/pull/13673/commits/772ff3e8cc191aeed6b6517db300ccbd0b52a23e

This change allows us to handle the null case when importing. In the end, I didn't need to use `allowDroppedNullPlaceholders`, so yeah, just need to not use the strict mode, so a JSON value would also be parsed successfully, not only objects and arrays. We can move this to a new PR and drop it from here, if you think is the right direction to go.

Reply:
> What I mean is that using the jsonParseStrict() currently requires the input JSON to be either an array or an object value, and apparently null is not, or I'm missing something.

Wait, so you're saying that basically jsoncpp can't handle JSON input consisting only of simple values? That's a weird design choice.

But no matter, fortunately `null` is the only value we care about so we could handle it even by simply checking if the file contains only that string.

> I don't get it. What is the empty assembly in this case? Like, nothing is generated by the compiler, so there is no assembly there.

There's no assembly. This could be handled just at the CLI level by skipping the whole import and just assuming empty outputs. Or we could make it possible to have zero files in `EVMAssemblyStack`, but that's probably too much hassle.

In any case, the point is that the behavior should be the same as when you give the compiler sources and they happen to contain no deployable contracts.

Reply:
> > What I mean is that using the jsonParseStrict() currently requires the input JSON to be either an array or an object value, and apparently null is not, or I'm missing something.
> 
> Wait, so you're saying that basically jsoncpp can't handle JSON input consisting only of simple values? That's a weird design choice.

No, it can. But in strict mode, which is what we are currently using there, the `strictRoot` setting is set to `true`, which means that the root object must be either an array or an object value, and `Json::null` is none of those, so the parse fails. See: https://open-source-parsers.github.io/jsoncpp-docs/doxygen/class_json_1_1_features.html#ae23176c14b2e79e81fb61fb1a8ab58ee

> But no matter, fortunately `null` is the only value we care about so we could handle it even by simply checking if the file contains only that string.

Yeah, I thought about that as well, but would prefer to rather handle it on the JSON level, but I'm fine with either.

> There's no assembly. This could be handled just at the CLI level by skipping the whole import and just assuming empty outputs. Or we could make it possible to have zero files in `EVMAssemblyStack`, but that's probably too much hassle.
> 
> In any case, the point is that the behavior should be the same as when you give the compiler sources and they happen to contain no deployable contracts.

Got it. Should I do this change in this PR then, or just create another one handling it in the CLI?

Reply:
This is not critical for the release so it's fine to do it separately. Does not have to go into this PR. Especially since it looks like it's not the only thing that should be corrected. There's also https://github.com/ethereum/solidity/pull/13673/files#r1196589412, which I'd put in a separate PR too.

Reply:
> This is not critical for the release so it's fine to do it separately. Does not have to go into this PR. Especially since it looks like it's not the only thing that should be corrected.

Right.

> There's also https://github.com/ethereum/solidity/pull/13673/files#r1196589412, which I'd put in a separate PR too.

Do you mean the CI change for parallel run? Just to be sure if the link is correct, because it seems to not show a specific file to me.

Reply:
No, no, not the one for parallel run :) That one definitely should stay. It was [Storing all possible subpaths is an overkill](https://github.com/ethereum/solidity/pull/13673#discussion_r1367342411)